Centralize logging

mpenick opened the pull request:
**What this PR does**:

In the previous setup each bundle initialized slf4j (and in turn,
initialized logback). This causes issues when using stateful appenders,
more specifically `ch.qos.logback.core.rolling.RollingFileAppender`.
Each bundle initialized its own copy of the appenders causing issues
with rolling over the log files.

This centralizes logging initialization to the `core` bundle which
exports the slf4j APIs. This results in the `core` bundle initializing
logback once and sharing a single instance of the appenders and
log configuration with all other bundles.

**Checklist**
- [X] Changes manually tested
- [x] Automated Tests added/updated

mpenick commented:
Tested using this configuration: (Note: `<maxFileSize>` is quite low to force log roll-over)

```xml
<?xml version="1.0" encoding="UTF-8"?>
<configuration scan="true" scanPeriod="60 seconds">
  <jmxConfigurator/>

  <appender name="SYSTEMLOG" class="ch.qos.logback.core.rolling.RollingFileAppender">
    <filter class="ch.qos.logback.classic.filter.ThresholdFilter">
      <level>INFO</level>
    </filter>
    <file>log/system.log</file>
    <rollingPolicy class="ch.qos.logback.core.rolling.SizeAndTimeBasedRollingPolicy">
      <!-- rollover daily -->
      <fileNamePattern>log/system.log.%d{yyyy-MM-dd}.%i.zip</fileNamePattern>
      <maxFileSize>2KB</maxFileSize>
      <maxHistory>7</maxHistory>
      <totalSizeCap>5GB</totalSizeCap>
    </rollingPolicy>
    <encoder>
      <pattern>%-5level [%thread] %date{ISO8601} %F:%L - %msg%n</pattern>
    </encoder>
  </appender>

  <!-- DEBUGLOG rolling file appender to debug.log (all levels) -->

  <appender name="DEBUGLOG" class="ch.qos.logback.core.rolling.RollingFileAppender">
    <file>log/debug.log</file>
    <rollingPolicy class="ch.qos.logback.core.rolling.SizeAndTimeBasedRollingPolicy">
      <!-- rollover daily -->
      <fileNamePattern>log/debug.log.%d{yyyy-MM-dd}.%i.zip</fileNamePattern>
      <maxFileSize>2KB</maxFileSize>
      <maxHistory>7</maxHistory>
      <totalSizeCap>5GB</totalSizeCap>
    </rollingPolicy>
    <encoder>
      <pattern>%-5level [%thread] %date{ISO8601} %F:%L - %msg%n</pattern>
    </encoder>
  </appender>

  <appender name="ASYNCDEBUGLOG" class="ch.qos.logback.classic.AsyncAppender">
    <queueSize>1024</queueSize>
    <discardingThreshold>0</discardingThreshold>
    <includeCallerData>true</includeCallerData>
    <appender-ref ref="DEBUGLOG" />
  </appender>

  <appender name="STDOUT" class="ch.qos.logback.core.ConsoleAppender">
    <encoder>
      <pattern>%-5level [%thread] %date{ISO8601} %F:%L - %msg%n</pattern>
    </encoder>
  </appender>

  <root level="${stargate.logging.level.root:-INFO}">
    <appender-ref ref="SYSTEMLOG" />
    <appender-ref ref="STDOUT" />
    <appender-ref ref="ASYNCDEBUGLOG" /> <!-- Comment this line to disable debug.log -->
  </root>

  <logger name="org.apache.cassandra" level="${stargate.logging.level.cassandra:-INFO}"/>
  <logger name="com.datastax.bdp.db" level="${stargate.logging.level.dse:-INFO}"/>
  <logger name="io.stargate.web" level="${stargate.logging.level.web:-INFO}"/>
  <logger name="io.stargate.db" level="${stargate.logging.level.persistence:-INFO}"/>
  <logger name="io.stargate.auth" level="${stargate.logging.level.auth:-INFO}"/>
</configuration>
```

mpenick commented:
Also, manually verified that log framework initialization only occurs once by adding breakpoints to, `org.slf4j.LoggerFactory#getILoggerFactory()` (this uses static init fields) and `ch.qos.logback.core.rolling. RollingFileAppender<E>#start()` to verify initialization only occurs once.

mpenick commented on auth-api/pom.xml:
```diff
@@ -151,9 +143,9 @@
-                            !org.slf4j,
-                            !org.slf4j.helpers,
-                            !org.slf4j.spi,
+                            org.slf4j,
```
@dougwettlaufer do you remember why this explicitly excluded `org.slf4j` before?

dougwettlaufer replied:
No but it's probably outdated now. As long as you can make a few requests to it and logs show up as expected then you should be good. I _think_ the big one is to make sure stacktraces show up because I vaguely remember there being some issue with those before.

ivansenic commented on core/pom.xml:
```diff
@@ -12,6 +12,7 @@
         <dependency>
             <groupId>org.slf4j</groupId>
             <artifactId>slf4j-api</artifactId>
+            <scope>provided</scope>
```
wait who will provide this now? if core is about to export this, it needs to have it on the runtime classpath?

mpenick replied:
nice catch. this was unintentional.

ivansenic commented on auth-api/pom.xml:
```diff
@@ -151,9 +143,9 @@
+                            org.slf4j,
+                            org.slf4j.helpers,
+                            org.slf4j.spi,
                             !ch.qos.logback.classic.*,
                             !ch.qos.logback.core.*,
```
This can be removed now as well.. As `logback-classic` is not anymore on the classpath this becomes obsolete..

ivansenic commented on auth-table-based-service/pom.xml:
```diff
@@ -42,17 +42,10 @@
-      <groupId>ch.qos.logback</groupId>
-      <artifactId>logback-core</artifactId>
-    </dependency>
     <dependency>
       <groupId>org.slf4j</groupId>
       <artifactId>slf4j-api</artifactId>
+      <scope>provided</scope>
     </dependency>
```
Can be removed and should be pulled with the `core` module.

ivansenic commented on core/pom.xml:
```diff
@@ -80,11 +81,35 @@
+                          org.slf4j.event,
+                          org.slf4j.helpers,
+                          org.slf4j.spi,
+                          ch.qos.logback.classic,
+                          ch.qos.logback.classic.helpers,
+                          ch.qos.logback.classic.spi,
```
Is there a need to export this at all if nobody is directly using and depending on the `logback`?

EDIT: OK I see now persistence is importing.. What is the reason? Is persistence setting up something in logback?

mpenick replied:
I address  this in the comment above.

mpenick replied:
Example of `cassandra-all` using logback: https://github.com/apache/cassandra/blob/cassandra-3.11/src/java/org/apache/cassandra/utils/logging/LogbackLoggingSupport.java#L69-L70

ivansenic replied:
OK was not aware of the `cassandra-all` using it..

ivansenic commented on graphqlapi/pom.xml:
```diff
@@ -115,20 +115,13 @@
     <dependency>
       <groupId>org.slf4j</groupId>
       <artifactId>slf4j-api</artifactId>
+      <scope>provided</scope>
     </dependency>
```
same here, can be removed and will be pulled with the `core` module..

ivansenic commented on auth-jwt-service/pom.xml:
```diff
@@ -26,15 +26,6 @@
-      <groupId>ch.qos.logback</groupId>
-      <artifactId>logback-core</artifactId>
-    </dependency>
     <dependency>
       <groupId>org.slf4j</groupId>
       <artifactId>slf4j-api</artifactId>
```
should this also be removed?

ivansenic replied:
Also do we really need here anymore:

```
    <dependency>
      <groupId>org.slf4j</groupId>
      <artifactId>log4j-over-slf4j</artifactId>
      <scope>runtime</scope>
    </dependency>
```

ivansenic commented on grpc/pom.xml:
```diff
@@ -114,6 +114,9 @@
             <Bundle-SymbolicName>io.stargate.grpc</Bundle-SymbolicName>
             <Bundle-Activator>io.stargate.grpc.GrpcActivator</Bundle-Activator>
             <Import-Package><![CDATA[
+                            org.slf4j,
```
dependency still there -> https://github.com/stargate/stargate/pull/1410/files#diff-45944c86e0452447902c811f67bc39b759eb37f3c675ef5c7fa291787124691dR62-R66

ivansenic reviewed (changes requested):
I am requesting changes because I think some stuff should be revisited and I want to have another look :nerd_face: 

* core should declare the `slf4j-api` as compile and runtime dependency (default scope) - and I am quite puzzled here how it worked until now.. 
* any module depending on core does not have to specify `slf4j-api` at all as dependency
* any module NOT depending on core has to specify `slf4j-api` as a provided dependency
* same applies to `logback` if export / import is needed at all? not sure why persistence modules depend on importing logback classes

mpenick commented:
> I am requesting changes because I think some stuff should be revisited and I want to have another look nerd_face
> 
> * core should declare the `slf4j-api` as compile and runtime dependency (default scope) - and I am quite puzzled here how it worked until now..

Not sure, but unintentional, fixed.

> * any module depending on core does not have to specify `slf4j-api` at all as dependency

Fixed.

> * any module NOT depending on core has to specify `slf4j-api` as a provided dependency

I'm not sure if this is true if the module doesn't use slf4j logging APIs.

> * same applies to `logback` if export / import is needed at all? not sure why persistence modules depend on importing logback classes

Discussed here: https://github.com/stargate/stargate/blob/2a40385fb6fce262aee5e467bf51527a7344c5a7/core/pom.xml#L90-L95.

ivansenic reviewed (approved):
Yes this is it! :+1:  

Hope that tests are still green :crossed_fingers:

ivansenic commented:
@mpenick about your comment:

> Ideally, this would not export `ch.qos.logback` APIs; however, the
                        persistence backends include dependencies (`cassandra-all` and `dse-db`)
                        that use logback APIs directly e.g. calling
                        `org.slf4j.LoggerFactory.getLogger(...)` and casting it to a logback
                        `ch.qos.logback.classic.Logger`. This results in casting errors without
                        exporting these APIs. Only the persistence backends import these.

So @tomekl007 got this error in the tests:

```
INFO  [Exec Stream Pumper] 2021-11-18 10:54:12,662 ProcessRunner.java:78 - stargate0-0> ERROR [StorageServiceShutdownHook] 2021-11-18 10:54:12,661 CassandraDaemon.java:244 - Exception in thread Thread[StorageServiceShutdownHook,5,main]
INFO  [Exec Stream Pumper] 2021-11-18 10:54:12,662 ProcessRunner.java:78 - stargate0-0> java.lang.IncompatibleClassChangeError: Class ch.qos.logback.classic.LoggerContext does not implement the requested interface ch.qos.logback.core.Context
INFO  [Exec Stream Pumper] 2021-11-18 10:54:12,662 ProcessRunner.java:78 - stargate0-0> 	at ch.qos.logback.core.spi.ContextAwareBase.addStatus(ContextAwareBase.java:78)
INFO  [Exec Stream Pumper] 2021-11-18 10:54:12,662 ProcessRunner.java:78 - stargate0-0> 	at ch.qos.logback.core.spi.ContextAwareBase.addInfo(ContextAwareBase.java:85)
INFO  [Exec Stream Pumper] 2021-11-18 10:54:12,662 ProcessRunner.java:78 - stargate0-0> 	at ch.qos.logback.core.hook.DelayingShutdownHook.run(DelayingShutdownHook.java:52)
INFO  [Exec Stream Pumper] 2021-11-18 10:54:12,662 ProcessRunner.java:78 - stargate0-0> 	at org.apache.cassandra.utils.logging.LogbackLoggingSupport.onShutdown(LogbackLoggingSupport.java:90)
INFO  [Exec Stream Pumper] 2021-11-18 10:54:12,662 ProcessRunner.java:78 - stargate0-0> 	at org.apache.cassandra.service.StorageService$1.runMayThrow(StorageService.java:701)
INFO  [Exec Stream Pumper] 2021-11-18 10:54:12,662 ProcessRunner.java:78 - stargate0-0> 	at org.apache.cassandra.utils.WrappedRunnable.run(WrappedRunnable.java:28)
INFO  [Exec Stream Pumper] 2021-11-18 10:54:12,662 ProcessRunner.java:78 - stargate0-0> 	at org.apache.cassandra.concurrent.NamedThreadFactory.lambda$threadLocalDeallocator$0(NamedThreadFactory.java:84)
INFO  [Exec Stream Pumper] 2021-11-18 10:54:12,662 ProcessRunner.java:78 - stargate0-0> 	at java.lang.Thread.run(Thread.java:748)
```

Seems very much like the problem with the persistence APIs, maybe all modules should also import the `ch.qos.logback.classic` stuff?

tomekl007 commented:
Ok, It may be a false alarm. I've rebuilt it again and the test is passing. However, I am still getting this error for this change:
https://github.com/stargate/stargate/compare/fix_grpc_connections_close?expand=1 
It might be related, but does not seem like a blocker.
